Jenkinsfile

```diff
@@ -0,0 +1,259 @@
+    dockerScan = true
+    cron = '@nightly'
+}
+
+def defaultParams = [
+    booleanParam(name: 'RELEASE_BUILD',
+        description: 'Is this a release build.',
+        defaultValue: false),
+    string(name: 'GIT_REVISION',
+        description: 'The git SHA to create the release build from.',
+        defaultValue: ''),
+    booleanParam(name: 'PROMOTE_TO_PRODUCTION',
+        defaultValue: false,
+        description: 'Promote images to production (DockerHub) for release build images only.'),
+    string(name: 'KSQLDB_VERSION',
+        defaultValue: '',
+        description: 'KSQLDB version to promote to production.'),
+    string(name: 'IMAGE_REVISION',
+        defaultValue: '1',
+        description: 'Revision for Docker images. This is used with PROMOTE_TO_PRODUCTION only'),
+    booleanParam(name: 'UPDATE_LATEST_TAG',
+        defaultValue: false,
+        description: 'Should the latest tag on docker hub be updated to point at this new image version.')
+]
+
+def updateConfig = { c ->
+    c.properties = [parameters(defaultParams)]
+}
+
+def finalConfig = jobConfig(config, [:], updateConfig)
+
+def job = {
+    if (params.RELEASE_BUILD && params.PROMOTE_TO_PRODUCTION) {
+        currentBuild.result = 'ABORTED'
+        error("You can not do a release build and promote to production at the same time.")
+    }
+
+    if (params.RELEASE_BUILD && params.GIT_REVISION == '') {
+        currentBuild.result = 'ABORTED'
+        error("You must provide the GIT_REVISION when doing a release build.")
+    }
+
+    if (params.PROMOTE_TO_PRODUCTION && params.KSQLDB_VERSION == '') {
+        currentBuild.result = 'ABORTED'
+        error("You must provide the KSQLDB_VERSION when promoting images to docker hub.")
+    }
+
+    config.release = params.RELEASE_BUILD
+
+    if (config.release) {
+        // For a release build check out the provided git sha instead of the master branch.
+        if (config.revision == '') {
+            error("If you are doing a release build you must provide a git sha.")
+        }
+        
+        config.revision = params.GIT_REVISION
+        // For a release build we remove the -SNAPSHOT from the version.
+        config.ksql_db_version = config.ksql_db_version.tokenize("-")[0]
+        config.docker_tag = config.ksql_db_version
+    } else {
+        config.docker_tag = config.ksql_db_version.tokenize("-")[0] + '-' + env.BUILD_NUMBER
+        config.revision = 'refs/heads/master'
+    }
+
+    if (params.PROMOTE_TO_PRODUCTION) {
+        withDockerServer([uri: dockerHost()]) {
+            config.dockerRepos.each { dockerRepo ->
+                sh "docker pull ${config.dockerRegistry}${dockerRepo}:${params.KSQLDB_VERSION}"
+                sh "docker tag ${config.dockerRegistry}${dockerRepo}:${params.KSQLDB_VERSION} ${dockerRepo}:${params.KSQLDB_VERSION}-${params.IMAGE_REVISION}"
+                sh "docker push ${dockerRepo}:${params.KSQLDB_VERSION}-${params.IMAGE_REVISION}"
```

Comment:
can we check and fail if this image already exists? I can also do this in a follow-up PR if you prefer

Reply:
lets do a follow up PR either you or me.